### ISSUE TITLE
Knex Configuration File and DotEnv package issue

### ISSUE THREAD

**Issue by devaku** [knex/knex#3939]
# Environment

Knex version: 
"knex": "^0.21.1",

Database + version:
"mysql": "^2.18.1",

OS: Windows 10

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do:

I was working with knex, using the dotenv package ("dotenv": "^8.2.0") and I ran across an odd issue.

Upon filling the knexconfig file with the following details, I was given the following error:
```
  development: {
        client: 'mysql',
        connection: {
            host: process.env.DB_HOST,
            user: process.env.DB_USERNAME,
            password: process.env.DB_PASSWORD,
            database: process.env.DB_NAME,
        },

```
```
Error: ER_ACCESS_DENIED_ERROR: Access denied for user ''@'localhost' (using password: NO)
    at Handshake.Sequence._packetToError (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\sequences\Sequence.js:47:14)
    at Handshake.ErrorPacket (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\sequences\Handshake.js:123:18)
    at Protocol._parsePacket (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\Protocol.js:291:23)
    at Parser._parsePacket (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\Parser.js:433:10)
    at Parser.write (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\Parser.js:43:10)
    at Protocol.write (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\Protocol.js:38:16)
    at Socket.<anonymous> (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\Connection.js:88:28)
    at Socket.<anonymous> (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\Connection.js:526:10)
    at Socket.emit (events.js:198:13)
    at addChunk (_stream_readable.js:288:12)
    --------------------
    at Protocol._enqueue (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\Protocol.js:144:48)
    at Protocol.handshake (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\Protocol.js:51:23)
    at Connection.connect (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\Connection.js:116:18)
    at Promise (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\knex\lib\dialects\mysql\index.js:69:18)
    at new Promise (<anonymous>)
    at Client_MySQL.acquireRawConnection (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\knex\lib\dialects\mysql\index.js:64:12)
    at create (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\knex\lib\client.js:291:39)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

At first I thought there was a credentials issue, but after determining there wasn't, I managed to narrow down the problem to this.

It would seem that for some reason, knex can't read the files that process.env is giving it. If I set the user object to 'root' (or specify a string directly into the config file) it works flawlessly.

Same goes for the database. If I set it using the dotenv package, I would get:

```
Error: ER_NO_DB_ERROR: No database selected
    at Query.Sequence._packetToError (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\sequences\Sequence.js:47:14)
    at Query.ErrorPacket (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\sequences\Query.js:79:18)
    at Protocol._parsePacket (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\Protocol.js:291:23)
    at Parser._parsePacket (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\Parser.js:433:10)
    at Parser.write (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\Parser.js:43:10)
    at Protocol.write (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\Protocol.js:38:16)
    at Socket.<anonymous> (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\Connection.js:88:28)
    at Socket.<anonymous> (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\Connection.js:526:10)
    at Socket.emit (events.js:198:13)
    at addChunk (_stream_readable.js:288:12)
    --------------------
    at Protocol._enqueue (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\protocol\Protocol.js:144:48)
    at Connection.query (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\mysql\lib\Connection.js:198:25)
    at C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\knex\lib\dialects\mysql\index.js:135:18
    at new Promise (<anonymous>)
    at Client_MySQL._query (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\knex\lib\dialects\mysql\index.js:129:12)
    at Client_MySQL.query (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\knex\lib\client.js:169:17)
    at Runner.query (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\knex\lib\runner.js:134:36)
    at Runner.queryArray (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\knex\lib\runner.js:226:19)
    at C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\knex\lib\runner.js:38:25
    at client.acquireConnection.catch.then (C:\Users\Valeen\Desktop\Codes\Webdevelopment\Personal\nodejs-knex-database\node_modules\knex\lib\runner.js:260:24)
```

So it looks like this is the correct format the knex file accepts:

```
host: process.env.DB_HOST,
user: 'root',
password: process.env.DB_PASSWORD,
database: 'nodejs_knex_database',
```

I don't understand why it's throwing an error at the 'user' and the 'database' objects. Any help would be greatly appreciated.

**Comment by Naddiseo** [comment]
I could be way off, but without seeing the dotenv file, I would assume a typo somewhere, or possibly something to do with how dotenv reads the values. My thinking is that the database is reporting an empty string as the user, so the `process.env.DB_USERNAME` is possibly undefined /empty. You could try console logging `console.log(JSON.stringify(process.env))` and see if all the variables are expected.

**Comment by devaku** [comment]
DB_HOST='localhost'
DB_USERNAME='root'
DB_PASSWORD=''
DB_NAME='knex_database'

Pretty much run of the mill stuff.

And yes, I did log it, and it showed up as normal. It's completely baffling me at the moment.

**Comment by Naddiseo** [comment]
Very strange. 
A couple debugging theories:
- Could you try `127.0.0.1` instead of `localhost`
- Could you try without the environment config stuff
- Could you try without the Knexfile and use knex directly
- Could you try making a new database user/password and using that

**Comment by devaku** [comment]
- The host was never the issue. It reads DB_HOST fine.

- Hardcoding it to the knex file works.

- I'm not sure how to do that.

- These are the default credentials for MySQL

**Comment by Naddiseo** [comment]
Sorry, should have clarified better:
- re-reading the issue, this one probably wont help diagnose
- What I mean is, in the Knexfile, you're using  `development: {` (and presumably `production:{`, I'm wondering if using the values directly will work:
   ```javascript
       // Knexfile.js
      module.exports = {
         client:'mysql',
         connection: { user: process.env.DB_USER, /* .... */ }
     }
    ```
    If removing the production/development keys works, then it might be an issue with `NODE_ENV` or maybe [liftoff](https://github.com/js-cli/js-liftoff)?[^1]
- I'm assuming you have a Knexfile.js that directly exports the config `module.exports = { /* knex config here */ };`, I'm wondering if getting rid of that file, and putting the config directly into the `const knex = Knex({ /* knex config using process.env here */ })` will change anything. If it work now, then I would expect, again, that it's an issue with the config reading part.
- also one that upon re-reading I don't think will help diagnose the issue

If it still doesn't work, would you be able to create a testing repo that I could test with?

[^1] - I'm not sure how this works exactly, I just see in the documentation that the Knexfile.js config support is done with liftoff

**Comment by elhigu** [comment]
Doesn't look like knex issue, please provide complete reproduction example. Can be opened if reproduction is provided.

**Comment by devaku** [comment]
Here you go guys. I got caught up with personal matters but I managed to finally get off my ass and upload this repository with the aforementioned issues with dotenv.

Just run `npm install` and then run `npm run debug`. You'll be able to see nodemon outlining that there are proper credentials to be used for MySQL. Just be sure to have a XAMP running in the background to get this to work of course.

But yes, If you guys can help me with this weird issue, it'll be great.

https://github.com/devaku/nodejs-knex-database
